### PR TITLE
Alabama Itemized Deductions

### LINKED ISSUE #2184: Alabama Itemized Deductions

### PR DESCRIPTION
**Pull request by dnnnnn0626**
Fixes #2184

### Summary
📄💰🧮

This pull request adds the logic and parameters for itemized deductions in Alabama income tax. It creates new YAML files for each deduction type and a Python file for the itemized deductions variable. It also adds unit tests for the variable in the `policyengine_us/tests` folder.

> _`Alabama` code_
> _Adding deductions, tests, files_
> _Autumn of taxes_

### Walkthrough
*  Add a new variable for Alabama itemized deductions ([link](https://github.com/PolicyEngine/policyengine-us/pull/2977/files?diff=unified&w=0#diff-9705579e71081a846018aa4cf2d9c14f8647c36ac793b5f354534ab9999f56faR1-R19))
*  Implement the formula for the variable by adding the values of the deduction variables ([link](https://github.com/PolicyEngine/policyengine-us/pull/2977/files?diff=unified&w=0#diff-9705579e71081a846018aa4cf2d9c14f8647c36ac793b5f354534ab9999f56faR1-R19))
*  Define the deduction variables and their metadata in YAML files under `policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized` ([link](https://github.com/PolicyEngine/policyengine-us/pull/2977/files?diff=unified&w=0#diff-23cdeaffed65e8c56376198bd333c7ece78ea079ed2b9bfb81e3031d7c74ffdeL1-R17), [link](https://github.com/PolicyEngine/policyengine-us/pull/2977/files?diff=unified&w=0#diff-68d73040c742e943d38b733d37ed53d9aff18edd25942b866d59201e7dda0d13L1-R12), [link](https://github.com/PolicyEngine/policyengine-us/pull/2977/files?diff=unified&w=0#diff-3a4754603b17b7951a325946f64965c6cc4170da848ba43373188f5abe971511R1-R15), [link](https://github.com/PolicyEngine/policyengine-us/pull/2977/files?diff=unified&w=0#diff-352b2c2a9bcec807fcb6b7993be4176d044f2bfc72d48dccef6f3ee2529995ceL1-R13), [link](https://github.com/PolicyEngine/policyengine-us/pull/2977/files?diff=unified&w=0#diff-4d69fca10d475ea919fade7e70c35d118145868005d40173476778d3b81c63b6R1-R12), [link](https://github.com/PolicyEngine/policyengine-us/pull/2977/files?diff=unified&w=0#diff-cd1d9ca4688159b922fe31b9f74ad6bca76a948d96d167d42c7f8905d710821eL1-R13), [link](https://github.com/PolicyEngine/policyengine-us/pull/2977/files?diff=unified&w=0#diff-188e10b55a07c649ae28ac8c614f56dd65a40986a006ac7b0e2e4b9c0b4e21dfR1-R13), [link](https://github.com/PolicyEngine/policyengine-us/pull/2977/files?diff=unified&w=0#diff-c4509cd0ba025af2bc7c66025c26ccefbe861f888cffd34c1abe1d4b9a1c34dbL1-R13))
*  Add unit tests for the variable in `policyengine_us/tests/policy/baseline/gov/states/al/tax/income/deductions/al_itemized_deduction.yaml` ([link](https://github.com/PolicyEngine/policyengine-us/pull/2977/files?diff=unified&w=0#diff-578fa2d0ab7f10cddd32be429066c41d7437adb2cdc93bc2f554c5006ee83323L1-R30))

### CODE REVIEW

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  2021-01-01:
+    - medical_and_dental_expense
+    - real_estate_tax
+    - mortage_interest_deduction
+    - investment_intertest_deduction
+    - charitable_deduction
+    - work_related_expense
+metadata:
+  unit: variable
+  name: al_itemized_deductions
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized.yaml:12]:
```suggestion
```

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+values:
+  2021-01-01:
+    - medical_and_dental_expense
+    - real_estate_tax
+    - mortage_interest_deduction
+    - investment_intertest_deduction
+    - charitable_deduction
+    - work_related_expense
+metadata:
+  unit: variable
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized.yaml:11]:
```suggestion
  unit: list
```

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/medical_expense.yaml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+description:  Medical expenses over this percentage of AGI are deductible from taxable income.
+  2021-01-01: 0.04
+
+metadata:
+  label:  Medical expenses over this percentage of AGI are deductible from taxable income.
+  period: year
+  unit: /1
+  reference:
+    - title: 2022 Schedule A (Form 1040)
+      href: https://www.revenue.alabama.gov/ultraviewer/viewer/basic_viewer/index.html?form=2023/01/22f40schabdc_blk.pdf
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/medical_expense.yaml:10]: add pages to the pdf links

Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/medical_expense.yaml:10]:
```suggestion
      href: https://www.revenue.alabama.gov/ultraviewer/viewer/basic_viewer/index.html?form=2023/01/22f40schabdc_blk.pdf#page=...
```

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/medical_expense.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/medical_expense.yaml:1]: create a al_medical_expense_deduction.py where you calculate the expenses over 4% of adjusted gross incoem

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized.yaml [hunk 4]
```diff
@@ -0,0 +1,18 @@
+description: Itemized deductions allowed in Alabama income tax.
+values:
+  2021-01-01:
+    - medical_and_dental_expense
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized.yaml:4]:
```suggestion
    - al_medical_expense_deduction
```

**Review by dnnnnn0626 — COMMENTED** [review]
Hi Paval, could you take a look at the code

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/excluded_deductions.yaml [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  2021-01-01:
+    - casualty_loss_deduction
+    - long_term_care_premium_deduction
+
+
+metadata:
+  unit: list
+  label: Itemized deductions excluded from Alabama.
+  reference:
+    - title: 2022 Schedule A (Form 1040)
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/excluded_deductions.yaml:12]: please also use this source for all parameters (point to exact section of the law document where you can find the parameter) f.e. - here: (13) 
https://law.justia.com/codes/alabama/2021/title-40/chapter-18/article-1/section-40-18-15/

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/excluded_deductions.yaml [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+      href: https://www.revenue.alabama.gov/ultraviewer/viewer/basic_viewer/index.html?form=2023/01/22f40schabdc_blk.pdf
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/excluded_deductions.yaml:13]: need to add paged to all pdf links. f.e. here:
```suggestion
      href: https://www.revenue.alabama.gov/ultraviewer/viewer/basic_viewer/index.html?form=2023/01/22f40schabdc_blk.pdf#page=1
```

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/excluded_deductions.yaml [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    - title: 2022 Form 40 Alabama Income Tax Return
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/excluded_deductions.yaml:14]: need to add 2021 form as well to make sure that there were no changes

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml [hunk 8]
```diff
@@ -0,0 +1,20 @@
+description: Itemized deductions allowed in Alabama income tax.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml:1]:
```suggestion
description: Alabama allows for he following itemized deductions.
```

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    - mortage_interest_deduction
+    - investment_intertest_deduction
+    - charitable_deduction
+    - FICA
+    - self_employment
+
+
+metadata:
+  unit: list
+  name: al_itemized_deductions
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml:14]:
```suggestion
```

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  label: AL itemized tax-unit-level deductions
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml:15]:
```suggestion
  label: Alabama itemized deductions
```

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/medical_expense.yaml [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+    - title: 2022 Form 40 Alabama Income Tax Return
+      href: https://www.revenue.alabama.gov/ultraviewer/viewer/basic_viewer/index.html?form=2023/02/22f40.pdf
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/medical_expense.yaml:12]: please also use this source for all parameters (point to exact section of the law document where you can find the parameter) f.e. - here: (13)
https://law.justia.com/codes/alabama/2021/title-40/chapter-18/article-1/section-40-18-15/

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/excluded_deductions.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/excluded_deductions.yaml:1]: I dont think we need this - we are not using it

File: policyengine_us/variables/gov/states/al/tax/income/al_medical_deduction.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    definition_period = YEAR
+    reference = (
+        "https://www.revenue.alabama.gov/ultraviewer/viewer/basic_viewer/index.html?form=2023/01/22f40schabdc_blk.pdf"
+        "https://www.revenue.alabama.gov/ultraviewer/viewer/basic_viewer/index.html?form=2023/02/22f40.pdf"
+    )
+    defined_for = StateCode.AL
+
+    def formula(tax_unit, period, parameters):
+        p = parameters(period).gov.states.al.tax.income.deductions.itemized
+        medical_deduction = tax_unit("al_medical_dedcution", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/al/tax/income/al_medical_deduction.py:18]: this is a circular reference? 

    def formula(tax_unit, period, parameters):
        expense = add(tax_unit, period, ["medical_expense"])
        medical = parameters(period).gov. - add alabama parameter path to the 4% 
        medical_floor = medical.floor * tax_unit("positive_agi", period)
        return max_(0, expense - medical_floor)

File: policyengine_us/variables/gov/states/al/tax/income/al_medical_deduction.py [hunk 13]
```diff
@@ -0,0 +1,20 @@
+from policyengine_us.model_api import *
+
+
+class al_medical_deduction(Variable):
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/al/tax/income/al_medical_deduction.py:4]:
```suggestion
class al_medical_expense_deduction(Variable):
```

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/al/tax/income/al_medical_deduction.py:4]: rename file name

File: policyengine_us/variables/gov/states/al/tax/income/al_medical_deduction.py [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+    value_type = float
+    entity = TaxUnit
+    label = "Alabama Medical Deduction"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/al/tax/income/al_medical_deduction.py:7]:
```suggestion
    label = "Alabama medical expense deduction"
```

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml [hunk 15]
```diff
@@ -0,0 +1,20 @@
+description: Itemized deductions allowed in Alabama income tax.
+values:
+  2021-01-01:
+    - real_estate_tax
+    - mortage_interest_deduction
+    - investment_intertest_deduction
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml:6]: this variable does not exist

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml:6]: interest_deduction

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+    - charitable_deduction
+    - FICA
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml:8]: is all of fica deductible? if so use taxsim_tfica

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+    - self_employment
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml:9]: this is not a deduction, are all of the self employment taxes deductable?

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/work_related_expense.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/work_related_expense.yaml:1]: delete and use 
parameters/gov/irs/deductions/itemized/misc/floor.yaml

File: policyengine_us/variables/gov/states/al/tax/income/al_work_related_expense.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/al/tax/income/al_work_related_expense.py:1]: this is the al_miscellaneous_deduction

File: policyengine_us/variables/gov/states/al/tax/income/al_work_related_expense.py [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+    definition_period = YEAR
+    reference = (
+        "https://www.revenue.alabama.gov/ultraviewer/viewer/basic_viewer/index.html?form=2023/01/22f40schabdc_blk.pdf"
+        "https://www.revenue.alabama.gov/ultraviewer/viewer/basic_viewer/index.html?form=2023/02/22f40.pdf"
+    )
+    defined_for = StateCode.AL
+
+    def formula(tax_unit, period, parameters):
+        p = parameters(period).gov.states.al.tax.income.deductions.itemized
+        work_related = tax_unit("work_related_expense", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/al/tax/income/al_work_related_expense.py:18]:
```suggestion
        work_related = tax_unit("misc_deduction", period)
```

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/excluded_deductions.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/excluded_deductions.yaml:1]: need changelog_entry file entry

File: policyengine_us/tests/policy/baseline/gov/states/al/tax/income/deductions/al_itemized_deduction.yaml [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+    mortage_interest_deduction: 7
+    investment_intertest_deduction: 1
+    charitable_deduction: 9
+    FICA: 2
+    self_employment: 1
+    al_medical_deduction: 2
+    al_work_related_expense: 2
+    state_code: AL
+  output:
+    al_itemized_deductions: 3+7+1+9+2+1+2+2+3
```
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/al/tax/income/deductions/al_itemized_deduction.yaml:14]: need full integer outputs

File: policyengine_us/tests/policy/baseline/gov/states/al/tax/income/deductions/al_work_related_expense.yaml [hunk 20]
```diff
@@ -0,0 +1,14 @@
+- name: Alabama work related itemized deductions unit test 1
+  period: 2021
+  input:
+    work_related_expense: 2000
```
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/al/tax/income/deductions/al_work_related_expense.yaml:4]:
```suggestion
    work_related_expense: 2_000
```

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml (hunk 17 above)
Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml:9]: self_employment_tax

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml:9]: iitax

**Review by dnnnnn0626 — COMMENTED** [review]
fix changes

File: policyengine_us/variables/gov/states/al/tax/income/al_itemized_deduction.py [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+    label = "Alabama itemized deductions"
+    unit = USD
+    definition_period = YEAR
+    reference = (
+        "https://law.justia.com/codes/alabama/2021/title-40/chapter-18/article-1/section-40-18-15/"
+        "https://www.revenue.alabama.gov/ultraviewer/viewer/basic_viewer/index.html?form=2023/01/22f40schabdc_blk.pdf#page=1"
+    )
+    defined_for = StateCode.AL
+
+    def formula(tax_unit, period, parameters):
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/al/tax/income/al_itemized_deduction.py:16]: Exclude the medical and salt deductions with the following code:

 deductions = [
            deduction
            for deduction in p.itemized_deductions
            if deduction
            not in [
                "salt_deduction",
                "medical_expense_deduction",
            ]
        ]

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml:1]: add to the list: 

al_medical_expense_deduction
al_misc_deduction

File: changelog_entry.yaml [hunk 22]
```diff
@@ -0,0 +1,4 @@
+- bump: minor
+  changes:
+    added:
+    - Alabama Itemized Deductions
```
Comment (PavelMakarchuk) [inline changelog_entry.yaml:4]:
```suggestion
    - Alabama itemized deductions.
```

File: policyengine_us/variables/gov/states/al/tax/income/al_medical_expense_deduction.py [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+    definition_period = YEAR
+    reference = (
+        "https://law.justia.com/codes/alabama/2021/title-40/chapter-18/article-1/section-40-18-15/"
+        "https://www.revenue.alabama.gov/ultraviewer/viewer/basic_viewer/index.html?form=2023/01/22f40schabdc_blk.pdf#page=1"
+    )
+    defined_for = StateCode.AL
+
+    def formula(tax_unit, period, parameters):
+        expense = add(tax_unit, period, ["medical_expense"])
+        medical = parameters(period).gov.states.al.tax.income.deductions.itemized.medical_expense
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/al/tax/income/al_medical_expense_deduction.py:18]:
```suggestion
        p = parameters(period).gov.states.al.tax.income.deductions.itemized
```

File: policyengine_us/variables/gov/states/al/tax/income/al_medical_expense_deduction.py [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+        medical_floor = medical.floor * tax_unit("positive_agi", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/al/tax/income/al_medical_expense_deduction.py:19]:
```suggestion
        medical_floor = p.medical_expense * tax_unit("positive_agi", period)
```

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/medical_expense.yaml [hunk 25]
```diff
@@ -0,0 +1,13 @@
+description: Medical expenses over this percentage of AGI are deductible from taxable income.
+values:
+  2021-01-01: 0.04
+
+metadata:
+  label: Medical expenses deduction threshold
+  period: year
+  unit: /1
+  reference:
+    - title: AL law documentation for itemized deductions
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/medical_expense.yaml:10]:
```suggestion
    - title: Code of Alabama Section 40-18-15, - then add exact passage where this policy parameter (4%) can be found
```

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/medical_expense.yaml:10]: use this format for all legal code references

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/medical_expense.yaml [hunk 26]
```diff
@@ -0,0 +1,13 @@
...
+      href: https://law.justia.com/codes/alabama/2021/title-40/chapter-18/article-1/section-40-18-15/
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/medical_expense.yaml:11]:
```suggestion
      href: https://law.justia.com/codes/alabama/2022/title-40/chapter-18/article-1/section-40-18-15/
```

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/medical_expense.yaml:11]: use 2022 link - all legal code references

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml (hunk 9 above)
Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml:15]: why resolved?

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/medical_expense.yaml [hunk 27]
```diff
@@ -0,0 +1,13 @@
...
+    - title: 2022 Schedule A (Form 1040)
```
[the remark this answers is not in the document]
Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/medical_expense.yaml:12]: for all yaml files

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/misc_deduction.yaml [hunk 28]
```diff
@@ -0,0 +1,14 @@
+description: Work related expenses over this percentage of AGI are deductible from taxable income.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/misc_deduction.yaml:1]:
```suggestion
description: Alabama deducts work related expenses over this percentage of adjusted gross income.
```

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/misc_deduction.yaml [hunk 29]
```diff
@@ -0,0 +1,14 @@
...
+values:
+  2021-01-01: 0.02
+
+metadata:
+  label: Work-related expense deduction threshold
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/misc_deduction.yaml:6]:
```suggestion
  label: Alabama work-related expense deduction rate
```

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/misc_deduction.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/misc_deduction.yaml:1]: rename to work_related_expense_rate.yaml

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml:1]: rename to sources.yaml

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml [hunk 30]
```diff
@@ -0,0 +1,22 @@
...
+    - al_medical_expense_deduction
+    - al_misc_deduction
+    - al_mortgage_interest_deduction
+    - al_investment_interest_deduction
+
+
+metadata:
+  unit: list
+  name: al_itemized_deductions
+  label: Alabama itemized deductions
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/itemized.yaml:16]:
```suggestion
  label: Alabama itemized deductions sources
```

File: policyengine_us/variables/gov/states/al/tax/income/al_itemized_deductions.py [hunk 31]
```diff
@@ -0,0 +1,25 @@
...
+    def formula(tax_unit, period, parameters):
+        return (
+            add(tax_unit, period, ["real_estate_taxes"])
+            + add(tax_unit, period, ["charitable_deduction"])
+            + add(tax_unit, period, ["taxsim_tfica"])
+            + tax_unit("al_medical_expense_deduction", period)
+            + tax_unit("al_misc_deduction", period)
+            + tax_unit("al_investment_interest_deduction", period)
+            + tax_unit("al_mortgage_interest_deduction", period)
+        )
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/al/tax/income/al_itemized_deductions.py:25]:
```suggestion
    adds = "gov.states.al.tax.income.deductions.itemized.sources"
```

File: policyengine_us/variables/gov/states/al/tax/income/al_mortgage_interest_deduction.py [hunk 32]
```diff
@@ -0,0 +1,14 @@
...
+
+class al_mortgage_interest_deduction(Variable):
+    value_type = float
+    entity = TaxUnit
+    label = "Alabama mortgage interest deduction"
+    defined_for = StateCode.AL
+    unit = USD
+    definition_period = YEAR
+    reference = (
+        "https://law.justia.com/codes/alabama/2022/title-40/chapter-18/article-1/section-40-18-15/"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/al/tax/income/al_mortgage_interest_deduction.py:12]: add sections of document as a comment after the link

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/al/tax/income/al_mortgage_interest_deduction.py:12]: all yaml

File: policyengine_us/variables/gov/states/al/tax/income/al_itemized_deductions.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/al/tax/income/al_itemized_deductions.py:1]: need `al_deductions` file that takes either the standard or the itemized deduction

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/al/tax/income/al_itemized_deductions.py:1]: also add it to the `al_taxable_income` concept

File: policyengine_us/variables/gov/states/al/tax/income/al_taxable_income.py [hunk 33]
```diff
@@ -9,4 +9,13 @@ class al_taxable_income(Variable):
...
+    reference = (
+        "https://law.justia.com/codes/alabama/2022/title-40/chapter-18/article-1/section-40-18-15/"  # Code of Alabama Section 40-18-15
+        "https://www.revenue.alabama.gov/ultraviewer/viewer/basic_viewer/index.html?form=2023/01/22f40schabdc_blk.pdf#page=1"  # 2022 Schedule A (Form 1040)
+        "https://www.revenue.alabama.gov/ultraviewer/viewer/basic_viewer/index.html?form=2022/06/21f40schabdc_blk.pdf#page=1"  # 2021 Schedule A (Form 1040)
+    )
+
+    def formula(tax_unit, period, parameters):
+        al_ded = tax_unit("al_deduction", period)
+        al_agi = tax_unit("al_agi", period)
+        return al_agi - al_ded
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/al/tax/income/al_taxable_income.py:21]:
```suggestion
        return max_(al_agi - al_ded, 0)
```

File: policyengine_us/variables/gov/states/al/tax/income/al_deduction.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/al/tax/income/al_deduction.py:1]: make a folder called deductions and include all files but taxable income

File: policyengine_us/variables/gov/states/al/tax/income/al_itemized_deductions.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/al/tax/income/al_itemized_deductions.py:1]: in the deductions folder an itemized folder

File: policyengine_us/tests/policy/baseline/gov/states/al/tax/income/al_taxable_income.yaml [hunk 34]
```diff
@@ -0,0 +1,19 @@
...
+
+- name: Alabama taxable income unit test 2
+  period: 2021
+  absolute_error_margin: 0.01
+  input:
+    al_deduction: 3
+    al_agi: 4
+    state_code: AL
+  output:
+    al_taxable_income: 1.00
```
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/al/tax/income/al_taxable_income.yaml:19]:
```suggestion
    al_taxable_income: 1
```

File: policyengine_us/tests/policy/baseline/gov/states/al/tax/income/al_taxable_income.yaml [hunk 35]
```diff
@@ -0,0 +1,19 @@
+- name: Alabama taxable income unit test 2
+  period: 2021
+  absolute_error_margin: 0.01
+  input:
+    al_deduction: 2
+    al_agi: 4
+    state_code: AL
+  output:
+    al_taxable_income: 2.00
```
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/al/tax/income/al_taxable_income.yaml:9]:
```suggestion
    al_taxable_income: 2
```

File: policyengine_us/variables/gov/states/al/tax/income/al_deduction.py [hunk 36]
```diff
@@ -0,0 +1,23 @@
+from policyengine_us.model_api import *
+
+
+class al_deduction(Variable):
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/al/tax/income/al_deduction.py:4]:
```suggestion
class al_deductions(Variable):
```

File: policyengine_us/variables/gov/states/al/tax/income/al_deduction.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/al/tax/income/al_deduction.py:1]: rename file to al_deductions.py

File: policyengine_us/tests/policy/baseline/gov/states/al/tax/income/deductions/al_deduction.yaml
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/al/tax/income/deductions/al_deduction.yaml:1]: rename file to al_deductions.yaml

File: policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/sources.yaml [hunk 37]
```diff
@@ -0,0 +1,24 @@
...
+    - taxsim_tfica
+    - al_medical_expense_deduction
+    - al_misc_deduction
+    - al_mortgage_interest_deduction
+    - al_investment_interest_deduction
+
+
+metadata:
+  unit: list
+  name: al_itemized_deductions
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/al/tax/income/deductions/itemized/sources.yaml:15]:
```suggestion
```

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: policyengine_us/variables/gov/states/al/tax/income/deductions/al_federal_tax_deduction.py [hunk 38]
```diff
@@ -0,0 +1,11 @@
+from policyengine_us.model_api import *
+
+
+class al_federal_tax_deduction_deduction(Variable):
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/al/tax/income/deductions/al_federal_tax_deduction.py:4]:
```suggestion
class al_federal_tax_deduction(Variable):
```

**Review by MaxGhenis — APPROVED** [review]
Good to merge after this

**Pull request merged**